src/nvmet.c

```diff
@@ -833,209 +909,230 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
 	}
-
-	list = clean_scandir(list, hostno);
-
-	/* Remove NVMe Target interface port */
-	if ((port != INVALID_VALUE) && (hostno == 0)) {
-		sprintf(path, "%s/%d/subsystems/%s%s-%s", SYS_NVMET_PORTS, port, NQN_HDR_STR, hostname, device);
-		if (access(path, F_OK) == SUCCESS) {
-			rc = unlink(path);
-			if (rc != SUCCESS) {
-				msg = "Error. Unable to remove NQN from port. %s: unlink: %s";
-				print_error(msg, return_message, __func__, strerror(errno));
+	if (allowed_host_number == 0) {
+		if (port != INVALID_VALUE) {
+			/* Example of this path:
+		 	 * /sys/kernel/config/nvmet/ports/<PORT>/subsystems/nqn.2021-06.org.rapiddisk:ubuserver-rd0
+	 	 	 */
+			sprintf(path, "%s/%d/subsystems/%s", SYS_NVMET_PORTS, port, nqn);
+			if (access(path, F_OK) == SUCCESS) {
+				if (unlink(path) != SUCCESS) {
+					msg = "Error. Unable to remove link of NQN to port. %s: unlink: %s";
+					print_error(msg, return_message, __func__, strerror(errno));
+					return rc;
+				}
+			}
+		} else {
+			sprintf(path, "%s", SYS_NVMET_PORTS);
+			if ((port_number = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
+				msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
+				print_error(msg, return_message, path, __func__, strerror(errno));
 				return rc;
 			}
-			msg = "Block device %s has been unmapped from NVMe Target port %d.";
-			if (return_message == NULL) {
-				print_error(msg, return_message, device, port);
-			} else {
-				sprintf(json_message_temp, msg, device, port);
-				if (strlen(json_message) > 0) {
-					strcat(json_message, " ");
+			for (n = 0; n < port_number; n++) {
+				sprintf(path, "%s/%s", SYS_NVMET_PORTS, list[n]->d_name);
+				if (access(path, F_OK) == SUCCESS) {
+					sprintf(path, "%s/%s/subsystems/%s", SYS_NVMET_PORTS, list[n]->d_name, nqn);
+					if (access(path, F_OK) == SUCCESS) {
+						if (unlink(path) != SUCCESS) {
+							msg = "Error. Unable to remove NQN %s. %s: rmdir: %s";
+							print_error(msg, return_message, path, __func__, strerror(errno));
+							list = clean_scandir(list, port_number);
+							return rc;
+						}
+					}
 				}
-				strcat(json_message, json_message_temp);
 			}
-		} else {
-			/* The NVMe Target interface port number that was defined does not exist on the system */
-			msg = "%s: Port %d and / or export does not exist.";
-			print_error(msg, return_message, __func__, port);
-			return rc;
+			list = clean_scandir(list, port_number);
 		}
-	}
-
-	/*
-	 * If there are no more allowed_hosts and the target is not being exported by the port, remove
-	 * the target export from the entire subsystem. NOTE - we are not checking other ports (yet).
-	 * This check will come in a future update.
-	 */
-	if ((port != INVALID_VALUE) && (hostno == 0)) {
-		if (access(path, F_OK) == SUCCESS) {
-			msg = "Error. Target is still exported through the port: %s.";
-			print_error(msg, return_message, __func__, path);
+		sprintf(path, "%s/%s/attr_allow_any_host", SYS_NVMET_TGT, nqn);
```

Comment:
Before we do this, we need to make sure that no other port number has this target NQN mapped in /sys/kernel/config/nvmet/ports/<PORT>/subsystems/. So, maybe another scandir and iterate through all ports.